### PR TITLE
Brokerage Integration - Samco (India 🇮🇳) 

### PR DESCRIPTION
**Pull request by rjra2611**
#### Description

Supports the following features for Samco Brokerage:

1. Enables subscription of DataFeeds
2. Live Trading
3. Historical Data Fetch
4. ToolBox Downloader

#### Related Issue

- Rebase form #5199
#### Motivation and Context

Support Indian Traders with Samco Brokerage Option 

#### Requires Documentation Change

- Local Live Trading: https://www.quantconnect.com/docs/v2/lean-cli/tutorials/live-trading/local-live-trading
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.CSharp/ZerodhaBasicTemplateOptionsAlgorithm.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+{
+    /// <summary>
+    /// This example demonstrates how to add options for a given underlying equity security.
+    /// It also shows how you can prefilter contracts easily based on strikes and expirations, and how you
+    /// can inspect the option chain to pick a specific option contract to trade.
+    /// </summary>
+    /// <meta name="tag" content="using data" />
+    /// <meta name="tag" content="options" />
+    /// <meta name="tag" content="filter selection" />
+    public class ZerodhaBasicTemplateOptionsAlgorithm : QCAlgorithm
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/ZerodhaBasicTemplateOptionsAlgorithm.cs:36]: `ZerodhaBasicTemplateOptionsAlgorithm -> should be named Samco?

File: Algorithm.CSharp/ZerodhaBasicTemplateOptionsAlgorithm.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+        /// <summary>
+        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
+        /// </summary>
+        public bool CanRunLocally { get; } = true;
+
+        /// <summary>
+        /// This is used by the regression test system to indicate which languages this algorithm is written in.
+        /// </summary>
+        public Language[] Languages { get; } = { Language.CSharp, Language.Python };
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/ZerodhaBasicTemplateOptionsAlgorithm.cs:113]: Are these values right?

File: Algorithm/QCAlgorithm.cs [hunk 3]
```diff
@@ -1561,7 +1561,7 @@ public Option AddOption(string underlying, Resolution? resolution = null, string
         /// <param name="leverage">The requested leverage for the </param>
         /// <returns></returns>
         /// <exception cref="KeyNotFoundException"></exception>
-        public Option AddOption(Symbol underlying, Resolution? resolution = null, string market = null, bool fillDataForward = true, decimal leverage = Security.NullLeverage)
+        public Option AddOption(Symbol underlying, Resolution? resolution = null, string market = null, bool fillDataForward = true, decimal leverage = Security.NullLeverage, OptionStyle optionStyle= OptionStyle.American)
```
Comment (Martin-Molinero) [inline Algorithm/QCAlgorithm.cs:1564]: `, OptionStyle optionStyle= OptionStyle.American)` not being used 👍

File: Brokerages/Samco/SamcoBrokerage.DataQueueHandler.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            return
+                (securityType == SecurityType.Equity ||
+                securityType == SecurityType.Option ||
+                securityType == SecurityType.Future) && (market == Market.India ||
+                market == Market.India ||
+                market == Market.India ||
+                market == Market.India ||
+                market == Market.India ||
+                market == Market.India ||
+                market == Market.India
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.DataQueueHandler.cs:94]: _Many_ `market == Market.India` ?

File: Brokerages/Samco/SamcoBrokerage.DataQueueUniverseProvider.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                default:
+                    lookupFunc = symbol => symbol.Value;
+                    break;
+            }
+
+
+            var symbols = new List<Symbol>();
+            foreach(var scripMaster in _symbolMapper.samcoTradableSymbolList)
+            {
+                symbols.Add(_symbolMapper.createLeanSymbol(scripMaster));
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.DataQueueUniverseProvider.cs:57]: This method should return the option/future chain for a given symbol, not sure that's what happening here? Same comment for the Zerodha version.,
If current `LookupSymbols` is not implemented/working, think adding support for `IDataQueueUniverseProvider` could be a second step PR for Samco and Zerodha

File: Brokerages/Samco/SamcoWebSocketClientWrapper.cs [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using QuantConnect.Logging;
+using QuantConnect.Util;
+
+namespace QuantConnect.Brokerages.Samco
+{
+    public class SamcoWebSocketClientWrapper : IWebSocket
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoWebSocketClientWrapper.cs:27]: We probably don't need this guy? Why not use the `WebSocketClientWrapper`

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 7]
```diff
@@ -0,0 +1,1084 @@
...
+                while (!_ctsFillMonitor.IsCancellationRequested)
+                {
+                    _fillMonitorResetEvent.WaitOne(TimeSpan.FromMilliseconds(_fillMonitorTimeout), _ctsFillMonitor.Token);
+
+                    foreach (var kvp in _pendingOrders)
+                    {
+                        var orderId = kvp.Key;
+                        var order = kvp.Value;
+
+                        var response = _samcoAPI.GetOrderDetails(orderId);
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:155]: Seems order events are not event driven, does Samco not have order events? Note Zeroodha does

File: Common/Orders/Fees/SamcoFeeModel.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+using System;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Orders.Fees
+{
+    /// <summary>
+    /// Provides the default implementation of <see cref="IFeeModel"/>
+    /// Refer to https://www.samco.in/technology/brokerage_calculator
+    /// </summary>
+    public class SamcoFeeModel : IFeeModel
```
Comment (Martin-Molinero) [inline Common/Orders/Fees/SamcoFeeModel.cs:25]: Seems this exactly the same to `ZerodhaFeeModel`, is this correct?

File: Common/Orders/SamcoOrderProperties.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// <summary>
+    /// Contains additional properties and settings for an order submitted to Zerodha Brokerage
+    /// </summary>
+    public class SamcoOrderProperties : ZerodhaOrderProperties
+    {
+        /// <summary>
+        /// Kite product type
+        /// </summary>
+        public string ProductType { get;}
```
Comment (Martin-Molinero) [inline Common/Orders/SamcoOrderProperties.cs:29]: If this is already in the base no need to define it again here?

File: Tests/Common/Packets/BacktestNodePacketTests.cs [hunk 10]
```diff
@@ -58,14 +58,14 @@ public void JobDatesAreRespected()
...
-                    { "Treynor Ratio", "-1.5"},
-                    { "Total Fees", "$7.00"} },
+                    { "Alpha", "0.209"},
+                    { "Beta", "-0.136"},
+                    { "Annual Standard Deviation", "0.272"},
+                    { "Annual Variance", "0.074"},
+                    { "Information Ratio", "0.018"},
+                    { "Tracking Error", "0.422"},
+                    { "Treynor Ratio", "-1.362"},
+                    { "Total Fees", "$6.62"} },
```
Comment (Martin-Molinero) [inline Tests/Common/Packets/BacktestNodePacketTests.cs:68]: Wouldn't expect these to change?

File: Tests/Brokerages/Samco/SamcoBrokerageTests.cs [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+
+namespace QuantConnect.Tests.Brokerages.Samco
+{
+    [TestFixture, Ignore("This test requires a configured and active Samco account")]
+    public class SamcoBrokerageTests
```
Comment (Martin-Molinero) [inline Tests/Brokerages/Samco/SamcoBrokerageTests.cs:23]: We have a base brokerage test setup at `BrokerageTests` Samco should inherit from it 👍, it can have additional tests too. Can follow other brokerages here 💯

File: Tests/Brokerages/Samco/SamcoBrokerageHistoryProviderTests.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using NUnit.Framework;
+using QuantConnect.Brokerages.Samco;
+using QuantConnect.Configuration;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Logging;
+using QuantConnect.Securities;
```
Comment (Martin-Molinero) [inline Tests/Brokerages/Samco/SamcoBrokerageHistoryProviderTests.cs:8]: nit - licenses missing here and bellow

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+                var castResolution = (Resolution)Enum.Parse(typeof(Resolution), resolution);
+                var castSecurityType = (SecurityType)Enum.Parse(typeof(SecurityType), securityType);
+
+                if (castSecurityType == SecurityType.Forex || castSecurityType == SecurityType.Cfd || castSecurityType == SecurityType.Crypto || castSecurityType == SecurityType.Base)
+                {
+                    throw new ArgumentException("Invalid security type: " + castSecurityType);
+                }
+
+                // Load settings from config.json and create downloader
+                var dataDirectory = Config.Get("data-directory", "D:\\SourceCode\\AlgoTrading\\Lean-1\\Data");
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:60]: nit - this guy shouldn't be here `D:\\SourceCode\\AlgoTrading\\Lean-1\\Data` 

Instead of this config.get could potentially use `Globals.DataFolder` directly?

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                    // Download data
+                    var pairObject = Symbol.Create(pair, castSecurityType, market);
+
+                    if (startDate >= endDate)
+                    {
+                        throw new ArgumentException("Invalid date range specified");
+                    }
+
+                    var start = startDate.ConvertTo(DateTimeZone.Utc, TimeZones.Kolkata);
+                    var end = endDate.ConvertTo(DateTimeZone.Utc, TimeZones.Kolkata);
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:73]: TZ hard coded should replace for MHDB entry information, similar to zerodha 👍. Same bellow should store the data in data time zone. We should double check in which TZ in Samco returning the data too

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 15]
```diff
@@ -0,0 +1,1084 @@
...
+                    OnOrderEvent(new OrderEvent(order,
+                        DateTime.UtcNow,
+                        OrderFee.Zero,
+                        "Samco Order Event")
+                    { Status = OrderStatus.Canceled });
+                }
+            }
+        }
+
+        private void EmitFillOrder(SamcoOrderResponse orderResponse)
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:319]: Following Lean code style, private methods at the bottom, first public methods

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 16]
```diff
@@ -0,0 +1,1084 @@
...
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Returns the brokerage account's base currency
+        /// </summary>
+        public override string AccountBaseCurrency => Currencies.INR;
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:392]: nit - properties above public methods 👍

**Review by Martin-Molinero — COMMENTED** [review]
Cool!
Leaving a few comments for initial review 👍

File: Brokerages/Samco/SamcoWebSocketClientWrapper.cs (hunk 6 above)
Comment (rjra2611) [reply Brokerages/Samco/SamcoWebSocketClientWrapper.cs:27]: We left this in Zerodha as well. Because of these brokerage specific logs

                            `catch (Exception e)
                            {
                                Log.Error(e, "Error in SamcoWebSocketClientWrapper connection task");
                            }

                            Log.Trace("SamcoWebSocketClientWrapper connection task ended.");`

File: Brokerages/Samco/SamcoBrokerageAPI.cs [hunk 17]
```diff
@@ -0,0 +1,460 @@
...
+                    Volume = candle.volume,
+                    Value = candle.close,
+                    DataType = MarketDataType.TradeBar,
+                    Period = Resolution.Minute.ToTimeSpan(),
+                    EndTime = candle.dateTime.AddMinutes(1)
+                };
+            }
+        }
+
+        public IEnumerable<TradeBar> GetHistoricalDailyCandles(string symbol, string exchange, DateTime startDateTime, DateTime endDateTime, Resolution resolution = Resolution.Daily)
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerageAPI.cs:144]: Unused?

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 18]
```diff
@@ -0,0 +1,1072 @@
...
+            _algorithm = algorithm;
+            _securityProvider = algorithm.Portfolio;
+            _aggregator = aggregator;
+            _samcoAPI = new SamcoBrokerageAPI();
+            _samcoAPI.Authorize(apiKey, apiSecret, yob);
+            _symbolMapper = new SamcoSymbolMapper();
+            _messageHandler = new BrokerageConcurrentMessageHandler<WebSocketMessage>(OnMessageImpl);
+
+            var subscriptionManager = new EventBasedDataQueueHandlerSubscriptionManager();
+            _algorithm.SetOptionChainProvider(new SamcoLiveOptionChainProvider(_symbolMapper));
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:118]: This doesn't seem right, why is this required? Can follow the pattern of other brokerages regarding this

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 19]
```diff
@@ -0,0 +1,1072 @@
...
+                    yield break;
+                }
+
+                foreach (var candle in candles.intradayCandleData)
+                {
+                    if (!exchange.DateTimeIsOpen(candle.dateTime) && !request.IncludeExtendedMarketHours)
+                    {
+                        continue;
+                    }
+                    yield return new TradeBar()
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:667]: We are always returning tradeBar data here, what if the history request if of QuoteBar type?

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 20]
```diff
@@ -0,0 +1,1072 @@
...
+                }
+            } while (latestTime < request.EndTimeUtc);
+        }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        public override void Dispose()
+        {
+        }
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:690]: At least have to dispose of the `_samcoAPI`, my guess is that there are probably some other component to dispose of, let's review it 👍

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 21]
```diff
@@ -0,0 +1,1072 @@
...
+                    catch (Exception exception)
+                    {
+                        Log.Error(exception);
+                        throw;
+                    }
+
+                }
+                var request = JsonConvert.SerializeObject(sub);
+                WebSocket.Send(request);
+                WebSocket.Send("\n");
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:788]: `WebSocket.Send("\n");` this doesn't seem right?

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 22]
```diff
@@ -0,0 +1,1072 @@
...
+        /// <param name="order">The order to be placed</param>
+        /// <returns>True if the request for a new order has been placed, false otherwise</returns>
+        public override bool PlaceOrder(Order order)
+        {
+            var submitted = false;
+
+            _messageHandler.WithLockedStream(() =>
+            {
+                var security = _securityProvider.GetSecurity(order.Symbol);
+                var orderFee = security.FeeModel.GetOrderFee(new OrderFeeParameters(security, order));
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:239]: The order fee should come with the filling not the submitting of the orders, right? same for order update

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 23]
```diff
@@ -0,0 +1,1072 @@
...
+                    submitted = true;
+                    return;
+                }
+
+                var message = $"Order failed, Order Id: {order.Id} timestamp: {order.Time} quantity: {order.Quantity} content: {orderResponse.statusMessage}";
+                OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Samco Order Event") { Status = OrderStatus.Invalid });
+                OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, -1, message));
+
+                submitted = true;
+                return;
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:369]: No need to return here really?

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+                            timeSpan = Time.OneHour;
+                            break;
+
+                        case Resolution.Daily:
+                            history = _samcoAPI.GetIntradayCandles(pairObject.ID.Symbol, exchange, start, end).ToList();
+                            timeSpan = Time.OneDay;
+                            break;
+                    }
+
+                    foreach (var bar in history)
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:107]: Believe `history` is already returning a TradeBar?

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+                            timeSpan = Time.OneMinute;
+                            break;
+
+                        case Resolution.Hour:
+                            history = _samcoAPI.GetIntradayCandles(pairObject.ID.Symbol, exchange, start, end).ToList();
+                            timeSpan = Time.OneHour;
+                            break;
+
+                        case Resolution.Daily:
+                            history = _samcoAPI.GetIntradayCandles(pairObject.ID.Symbol, exchange, start, end).ToList();
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:102]: `history = _samcoAPI.GetIntradayCandles(pairObject.ID.Symbol, exchange, start, end).ToList();` seems the same in all the cases of the switch? we can probably remove it from it?

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+                foreach (var pair in tickers)
+                {
+                    // Download data
+                    var pairObject = Symbol.Create(pair, castSecurityType, market);
+
+                    var exchangeTimeZone = marketHoursDatabase.GetExchangeHours(market, pairObject, castSecurityType).TimeZone;
+                    var exchange = symbolMapper.GetDefaultExchange(pairObject);
+
+                    var start = startDate.ConvertTo(DateTimeZone.Utc, exchangeTimeZone);
+                    var end = endDate.ConvertTo(DateTimeZone.Utc, exchangeTimeZone);
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:79]: Samco API expects start and end in UTC?

File: Algorithm.CSharp/SamcoBasicTemplateOptionsAlgorithm.cs [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+        public Symbol OptionSymbol;
+
+        public override void Initialize()
+        {
+            //SetStartDate(2015, 12, 24);
+            //SetEndDate(2015, 12, 24);
+            //SetCash(100000);
+
+            SetTimeZone(TimeZones.Kolkata);
+            SetBenchmark(t => 0);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/SamcoBasicTemplateOptionsAlgorithm.cs:48]: `SetBenchmark(t => 0);` is repeated twice - Believe the `Samco` brokerage model should override the default benchmark 👍 Same for the `ZerodhaBrokerageModel`

File: Algorithm.CSharp/SamcoBasicTemplateOptionsAlgorithm.cs [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+    public class SamcoBasicTemplateOptionsAlgorithm : QCAlgorithm
+    {
+        private const string UnderlyingTicker = "NIFTY";
+        public Symbol OptionSymbol;
+
+        public override void Initialize()
+        {
+            //SetStartDate(2015, 12, 24);
+            //SetEndDate(2015, 12, 24);
+            //SetCash(100000);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/SamcoBasicTemplateOptionsAlgorithm.cs:45]: Commented code 😱

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+                    // Write data
+                    var writer = new LeanDataWriter(castResolution, pairObject, dataDirectory);
+                    IList<TradeBar> fileEnum = new List<TradeBar>();
+                    var history = new List<TradeBar>();
+                    if (castResolution == Resolution.Tick)
+                    {
+                        throw new ArgumentException("Samco Doesn't support tick resolution");
+                    }
+
+                    history = _samcoAPI.GetIntradayCandles(pairObject, exchange, startDate, endDate, isIndex: isIndex).ToList();
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:86]: nit- don't need to call `ToList` here if going to iterate again bellow in the foreach

File: ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+                var marketHoursDatabase = MarketHoursDatabase.FromDataFolder();
+                var symbolMapper = new SamcoSymbolMapper();
+
+                foreach (var pair in tickers)
+                {
+                    // Download data
+                    var pairObject = Symbol.Create(pair, castSecurityType, market);
+                    var exchange = symbolMapper.GetDefaultExchange(pairObject);
+                    var scrip = symbolMapper.SamcoSymbols.Where(x => x.Name.ToUpperInvariant() == pairObject.ID.Symbol).First();
+                    var isIndex = scrip.Instrument == "INDEX";
```
Comment (Martin-Molinero) [inline ToolBox/SamcoDataDownloader/SamcoDataDownloader.cs:75]: if `isIndex` is true, should it be a Index security type from Leans perspective?

File: Brokerages/Samco/SamcoMessages.cs [hunk 31]
```diff
@@ -0,0 +1,474 @@
...
+        public long LosingTodayCount { get; set; }
+
+        [JsonProperty("totalGainAndLossAmount")]
+        public string TotalGainAndLossAmount { get; set; }
+
+        [JsonProperty("dayGainAndLossAmount")]
+        public string DayGainAndLossAmount { get; set; }
+    }
+
+    internal class ParseStringConverter : JsonConverter
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Brokerages/Samco/SamcoMessages.cs:447]: Ping, we can remove `ParseStringConverter` I think?

File: Algorithm.CSharp/SamcoBasicTemplateOptionsAlgorithm.cs [hunk 32]
```diff
@@ -0,0 +1,164 @@
...
+
+            // set our strike/expiry filter for this option chain
+            option.SetFilter(u => u.Strikes(-2, +2)
+                                   // Expiration method accepts TimeSpan objects or integer for
+                                   // days. The following statements yield the same filtering criteria
+                                   .Expiration(0, 180));
+            // .Expiration(TimeSpan.Zero, TimeSpan.FromDays(180)));
+
+            // use the underlying equity as the benchmark
+            SetBenchmark(t => 0);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/SamcoBasicTemplateOptionsAlgorithm.cs:58]: >  SetBenchmark(t => 0);

Think this should be removed now since the default benchmark issue is solved?

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 33]
```diff
@@ -0,0 +1,1009 @@
...
+                    }
+                }
+
+                var brokerageSecurityType = _symbolMapper.GetBrokerageSecurityType(orderDetails.tradingSymbol);
+                var symbol = _symbolMapper.GetLeanSymbol(orderDetails.tradingSymbol, brokerageSecurityType);
+                var fillPrice = decimal.Parse(orderDetails.filledPrice, NumberStyles.Float, CultureInfo.InvariantCulture);
+                var fillQuantity = decimal.Parse(orderDetails.filledQuantity, NumberStyles.Float, CultureInfo.InvariantCulture);
+                var updTime = DateTime.UtcNow;
+                var security = _securityProvider.GetSecurity(order.Symbol);
+                var orderFee = security.FeeModel.GetOrderFee(new OrderFeeParameters(security, order));
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:707]: Fees should be on the filled quantity only really, else for partial fills will apply multiple times the same fees

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 34]
```diff
@@ -635,7 +647,10 @@ public void Subscribe(IEnumerable<Symbol> symbols)
                 }
             }
             var request = JsonConvert.SerializeObject(sub);
+            // required to flush input json as per samco forum
+            request = request + "\n";
             WebSocket.Send(request);
+            Unsubscribe(symbols);
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:599]: Why is subscribe calling Unsubscribe?

File: Brokerages/Samco/SamcoBrokerage.cs [hunk 35]
```diff
@@ -613,19 +613,31 @@ public void Subscribe(IEnumerable<Symbol> symbols)
...
+            }
             foreach (var symbol in symbols)
             {
                 try
                 {
                     var quote = GetQuote(symbol);
-                    _subscriptionsById[quote.listingId] = symbol;
-                    sub.request.data.symbols.Add(new Subscription.Symbol { symbol = quote.listingId });
-                    if (!subscribeInstrumentTokens.Contains(quote.listingId))
+                    var listingId = quote.listingId;
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerage.cs:580]: Doesn't the symbol mapped have the `listingId` to provide?

File: Brokerages/Samco/SamcoBrokerageAPI.cs [hunk 36]
```diff
@@ -0,0 +1,448 @@
...
+                        Symbol = leanSymbol,
+                        Low = candle.low,
+                        High = candle.high,
+                        Open = candle.open,
+                        Close = candle.close,
+                        Volume = candle.volume,
+                        Value = candle.close,
+                        DataType = MarketDataType.TradeBar,
+                        Period = resolution.ToTimeSpan(),
+                        EndTime = candle.dateTime.AddMinutes(1)
```
Comment (Martin-Molinero) [inline Brokerages/Samco/SamcoBrokerageAPI.cs:222]: > .AddMinutes(1)

Would depend in from the resolution really right? Either way shouldn't need to set EndTime, it's created by Time + Period already

File: Brokerages/Samco/SamcoLiveOptionChainProvider.cs [hunk 37]
```diff
@@ -0,0 +1,55 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace QuantConnect.Brokerages.Samco
+{
+    /// <summary>
+    /// An implementation of <see cref="IOptionChainProvider"/> that fetches the list of contracts
+    /// from the Samco StockNote API
+    /// </summary>
+    class SamcoLiveOptionChainProvider : IOptionChainProvider
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Brokerages/Samco/SamcoLiveOptionChainProvider.cs:27]: Let's remove this guy for now 👍 he is private btw, can't be used really

File: Brokerages/WebSocketClientWrapper.cs [hunk 38]
```diff
@@ -41,9 +42,11 @@ public class WebSocketClientWrapper : IWebSocket
         /// Wraps constructor
         /// </summary>
         /// <param name="url"></param>
-        public void Initialize(string url)
+        /// <param name="sessionToken"></param>
+        public void Initialize(string url, string sessionToken = null)
```
[the remark this answers is not in the document]
Comment (rjra2611) [reply Brokerages/WebSocketClientWrapper.cs:46]: Added for both `url ` and `sessionToken`  🙂

**Comment by Martin-Molinero** [comment]
Looks good ✅
Closing this PR so core implementation is be moved to https://github.com/QuantConnect/Lean.Brokerages.Samco and a new reduced PR for Lean is opened
